File: Graphs/802. Find Eventual Safe States.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        vector<int> s;

        int n = graph.size();
        vector<int> dp(n, -1);
        vector<int> isVisiting(n, -1);
        for (int i = 0; i < n; i++) {
            if (dfs(i, graph, dp, isVisiting)) s.push_back(i);
        }
        return s;



    }

    int dfs(int node, vector<vector<int>>& graph, vector<int>& dp, vector<int>& isVisiting) {
        if (graph[node].size() == 0) return 1;
        if (isVisiting[node] != -1) return 0;
        if (dp[node] != -1) return dp[node];

        int can = 1;
        isVisiting[node] = 1;
        for (auto nei: graph[node]) {
            int res = dfs(nei, graph, dp, isVisiting);
            can &= res;
        }
        isVisiting[node] = -1;
        dp[node] = can;
        return can;
    }



};

File: Graphs/802. Find Eventual Safe States.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        res = []
        nodes = len(graph)
        cache = {}
        isVisiting = set()

        def dfs(node):
            if len(graph[node]) == 0:
                return True
            if node in isVisiting:
                return False
            if node in cache:
                return cache[node]

            can = True
            isVisiting.add(node)
            for nei in graph[node]:
                t = dfs(nei)
                can = can and t
            isVisiting.discard(node)
            cache[node] = can

            return can

        n = len(graph)
        for i in range(n):

            if dfs(i):
                res.append(i)
            # print(i)
            # print(cache)

        return res
